source/NVDAObjects/window/edit.py

```diff
@@ -340,37 +344,62 @@ def _getStoryLength(self):
+			# As unicode characters are two bytes in size,
+			# this ensures that our buffer can hold both ANSI and unicode character strings.
+			bufLen = ((end - start) + 1) * 2
+			# Even though this can return unicode text, we use the ANSI version of the structure.
+			# Using the unicode structure isn't strictly necessary and saves us some confusion
+			textRange = TextRangeAStruct()
+			textRange.chrg.cpMin = start
+			textRange.chrg.cpMax = end
+			processHandle = self.obj.processHandle
+			internalBuf = winKernel.virtualAllocEx(processHandle, None, bufLen, winKernel.MEM_COMMIT, winKernel.PAGE_READWRITE)
 			try:
-				textRange.lpstrText=internalBuf
-				internalTextRange=winKernel.virtualAllocEx(processHandle,None,ctypes.sizeof(textRange),winKernel.MEM_COMMIT,winKernel.PAGE_READWRITE)
+				textRange.lpstrText = internalBuf
+				internalTextRange = winKernel.virtualAllocEx(processHandle, None, ctypes.sizeof(textRange), winKernel.MEM_COMMIT, winKernel.PAGE_READWRITE)
 				try:
-					winKernel.writeProcessMemory(processHandle,internalTextRange,ctypes.byref(textRange),ctypes.sizeof(textRange),None)
-					res=watchdog.cancellableSendMessage(self.obj.windowHandle,EM_GETTEXTRANGE,0,internalTextRange)
+					winKernel.writeProcessMemory(processHandle, internalTextRange, ctypes.byref(textRange), ctypes.sizeof(textRange), None)
+					# EM_GETTEXTRANGE returns the number of characters copied,
+					# not including the terminating null character.
+					# See https://docs.microsoft.com/en-us/windows/desktop/controls/em-gettextrange
+					numChars = watchdog.cancellableSendMessage(self.obj.windowHandle, EM_GETTEXTRANGE, 0, internalTextRange)
 				finally:
-					winKernel.virtualFreeEx(processHandle,internalTextRange,0,winKernel.MEM_RELEASE)
-				buf=(ctypes.c_byte*bufLen)()
+					winKernel.virtualFreeEx(processHandle, internalTextRange, 0, winKernel.MEM_RELEASE)
+				buf = (ctypes.c_byte * bufLen)()
+				# Copy the text in the text range to our own buffer.
 				winKernel.readProcessMemory(processHandle,internalBuf,buf,bufLen,None)
 			finally:
-				winKernel.virtualFreeEx(processHandle,internalBuf,0,winKernel.MEM_RELEASE)
-			if self.obj.isWindowUnicode or (res>1 and (buf[res]!=0 or buf[res+1]!=0)): 
+				winKernel.virtualFreeEx(processHandle, internalBuf, 0, winKernel.MEM_RELEASE)
+			# Find out the byte size for the characters in the buffer.
+			if (
+				# The window is unicode, the text range contains multi byte characters.
+				self.obj.isWindowUnicode
+				# Or the window reports being ANSI but the buffer seems to contain unicode characters.
+				# ANSI strings are terminated with one NULL terminated character.
+				# As pointed out above, numChars contains the number of characters without the null terminator.
+				# Therefore, if the string we got isn't null terminated at numCHars or numChars + 1,
+				# the buffer definitely contains multibyte characters.

```

Comment:
I think there is something missing in the explanation, and I dont think this actually does what we expect.

If I am understanding right, it relies on the value of `numChars` being half of the number of bytes in a wide character string and it being impossible to have a `null` character (represented by two null bytes) at the mid-point of the wide character string. If it is an ANSI string, there MUST be at least one `null` character at `numChars+1` to indicate the end of the string. However note that it does not handle an odd number of chars well. For wide character strings it will be considering one byte from one character, and one byte from the next. 

Also worth point out this relies on the memory being zeroed by the `VirtualAllocEx` function, otherwise:
```
numChars = 2 #  copy logical "hi" into buffer
# imaging a string copied to a buffer with "garbage" in it. The string copied is "hi"
# wide string memory "0h0i" copied over garbage
bW = "0h0i00garbage" # ok, bW[numChars] == 0 but bW[numChars+1] == 'i' so it is interpreted as unicode
# ANSI string memory "hi" copied over garbage
bA = "hi0garbage" # not ok, bW[numChars] == 0 but bW[numChars+1] == 'g' so it is interpreted as unicode
```

I think there is a broader issue here that odd numbers of characters are not handled correctly.

Consider where a wide string with `numChars = 3`:
```
buf = b"\x00\x04\x41\x00\x00\x01\x00\x00" # = "ЀAĀ" followed by two null chars.
buf[numChars] != 0 # False, it is zero
buf[numChars+1 != 0 # False, it is zero: buf interpreted as ANSI
"""

Would it be clearer to do something like:
```
if (
	# The window is unicode, the text range contains multi byte characters.
	self.obj.isWindowUnicode
	# VirtualAllocEx zeroes out memory, so for an ANSI string there should ONLY be null bytes 
	# after the first numchars . 
	# For wide character strings, the number of bytes will be twice the number of characters, meaning 
	# numchars points to mid-string, there will be non-zero bytes after it.
	or any(c != 0 for buf[numChars:bufLen])
):
	text=ctypes.cast(buf,ctypes.c_wchar_p).value
else:
	encoding=locale.getlocale()[1]
```

Also worth considering a single char. 
```
buf = b"\x41\x00\x00\x00" # = "A" followed by two null chars. This is valid as a wide string or ANSI
buf = b"\x00\x45\x00\x00" # ??
```
There are enough cases it would be worth extracting this and creating a unit test.

Although for perf considerations perhaps we just consider adjusting the byte alignment and checking two bytes
```
wCharAlignedIndex = numChars+(numChars%2)
	or not (buf[wCharAlignedIndex ] == 0 and buf[wCharAlignedIndex + 1] == 0) 
```

Reply:
We discussed this by voice, and I think I've been able to come up with working implementations that prove to be working in Wordpad and Notepad++. Note that in both situations, I also tried to fetch a range that was actually longer than the text in the control, so these cases are also covered.

While I really see your point about unit tests, note that we can't test this against neither Wordpad nor Notepad++. I think we should also split this logic out to a new function in textUtils that takes a buffer, a number of bytes and an encoding. The more I think about that, the more I consider it sensible, so that's what I'll do.